### DISCUSSION TITLE
`BlockingPortalProvider` fails on 2nd call with `RuntimeError: Event loop is closed`

### DISCUSSION THREAD

**Discussion by dhirschfeld** [agronholm/anyio#743] [Q&A] [answered]
**`sync_wrapper.py`**
```python
from __future__ import annotations

import trio
import anyio
from ipaddress import IPv4Address

import httpx


from anyio.from_thread import BlockingPortalProvider



class AsyncAPI:
    _client: httpx.AsyncClient

    def __init__(self):
        self._client = httpx.AsyncClient()

    @property
    def sync(self) -> SyncAPI:
        return SyncAPI(self)

    async def get_ipaddress(self) -> IPv4Address:
        url = "https://ifconfig.me/ip"
        response = await self._client.get(url)
        response.raise_for_status()
        return IPv4Address(response.text)


class SyncAPI:
    _portal_provider: BlockingPortalProvider
    _async_api: AsyncAPI

    def __init__(self, api: AsyncAPI):
        self._portal_provider = BlockingPortalProvider()
        self._async_api = api

    def get_ipaddress(self) -> IPv4Address:
        with self._portal_provider as portal:
            return portal.call(self._async_api.get_ipaddress)


api = AsyncAPI()

print('#'*80)
print('# First Call')
print('#'*80)
ipaddress = api.sync.get_ipaddress()
print(repr(ipaddress))

print()
print('#'*80)
print('# Second Call')
print('#'*80)
ipaddress = api.sync.get_ipaddress()
print(repr(ipaddress))
```

```
❯ python sync_wrapper.py
################################################################################
# First Call
################################################################################
IPv4Address('<redacted>')

################################################################################
# Second Call
################################################################################
Traceback (most recent call last):
  File "~/code/sandbox/sync_wrapper.py", line 56, in <module>
    ipaddress = api.sync.get_ipaddress()
  File ~/code/sandbox/sync_wrapper.py", line 41, in get_ipaddress
    return portal.call(self._async_api.get_ipaddress)
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/anyio/from_thread.py", line 287, in call
    return cast(T_Retval, self.start_task_soon(func, *args).result())
  File "/opt/python/envs/dev310/lib/python3.10/concurrent/futures/_base.py", line 458, in result
    return self.__get_result()
  File "/opt/python/envs/dev310/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/anyio/from_thread.py", line 218, in _call_func
    retval = await retval_or_awaitable
  File "~/code/sandbox/sync_wrapper.py", line 26, in get_ipaddress
    response = await self._client.get(url)
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpx/_client.py", line 1801, in get
    return await self.request(
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpx/_client.py", line 1574, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpx/_client.py", line 1661, in send
    response = await self._send_handling_auth(
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpx/_client.py", line 1689, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpx/_client.py", line 1726, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpx/_client.py", line 1763, in _send_single_request
    response = await transport.handle_async_request(request)
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpx/_transports/default.py", line 373, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 216, in handle_async_request
    raise exc from None
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 196, in handle_async_request
    response = await connection.handle_async_request(
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpcore/_async/connection.py", line 101, in handle_async_request
    return await self._connection.handle_async_request(request)
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpcore/_async/http11.py", line 142, in handle_async_request
    await self._response_closed()
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpcore/_async/http11.py", line 257, in _response_closed
    await self.aclose()
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpcore/_async/http11.py", line 265, in aclose
    await self._network_stream.aclose()
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/httpcore/_backends/anyio.py", line 55, in aclose
    await self._stream.aclose()
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/anyio/streams/tls.py", line 202, in aclose
    await self.transport_stream.aclose()
  File "/opt/python/envs/dev310/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 1202, in aclose
    self._transport.close()
  File "/opt/python/envs/dev310/lib/python3.10/asyncio/selector_events.py", line 706, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/opt/python/envs/dev310/lib/python3.10/asyncio/base_events.py", line 753, in call_soon
    self._check_closed()
  File "/opt/python/envs/dev310/lib/python3.10/asyncio/base_events.py", line 515, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
```

**Comment by dhirschfeld** [comment]
I've opened a discussion as it's likely user error, but I'm not sure what I'm doing wrong above?

**Comment by dhirschfeld** [comment]
> [!NOTE]
> It does seem to work fine if I instead use `BlockingPortalProvider('trio')`

**Comment by agronholm** [comment]
Ideally, you should have the provider as a global, so that other instances of your classes can take advantage of it. If you have a private instance, then it may be pointless to use the provider vs just opening a regular blocking portal.

**Reply by agronholm** [reply]
Does this occur with any other third party library besides httpx?

**Comment by dhirschfeld** [comment]
AFAICT this usage mirrors the example in the docs:
```python
from anyio.to_thread import BlockingPortalProvider

class MyAPI:
    def __init__(self, async_obj) -> None:
        self._async_obj = async_obj
        self._portal_provider = BlockingPortalProvider()

    def do_stuff(self) -> None:
        with self._portal_provider as portal:
            portal.call(async_obj.do_async_stuff)
```
https://anyio.readthedocs.io/en/stable/threads.html#sharing-a-blocking-portal-on-demand

**Comment by dhirschfeld** [comment]
If I make `portal_provider` a module global it still results in the `Event loop is closed` error:
```python
portal_provider = BlockingPortalProvider('asyncio')

class SyncAPI:
    _async_api: AsyncAPI

    def __init__(self, api: AsyncAPI):
        self._async_api = api

    def get_ipaddress(self) -> IPv4Address:
        with portal_provider as portal:
            return portal.call(self._async_api.get_ipaddress)
```

**Comment by dhirschfeld** [comment]
storing a `BlockingPortal` instance on the class "fixes" the error with `asyncio`:
```python
class SyncAPI:
    _async_api: AsyncAPI

    def __init__(self, api: AsyncAPI):
        self._async_api = api
        self._portal = portal_provider.__enter__()

    def get_ipaddress(self) -> IPv4Address:
        return self._portal.call(self._async_api.get_ipaddress)
```
```python
>>> api.sync.get_ipaddress()
IPv4Address('<redacted>')
>>> api.sync.get_ipaddress()
IPv4Address('<redacted>')
>>> api.sync.get_ipaddress()
IPv4Address('<redacted>')
```

So, maybe I just need to manually create a `BlockingPortal` instance in the `__init__` and provide a `close` method which cleaned up the portal?

**Reply by dhirschfeld** [reply]
> So, maybe I just need to manually create a `BlockingPortal` instance in the `__init__` and provide a `close` method which cleaned up the portal?

That doesn't look entirely straightforward 🤔 

If I don't care about `asyncio` support can I just use the `trio` backend with the portal-on-demand api?

**Reply by agronholm** [reply]
This actually boils down to the lifecycle management of the event loop. The `smtpproto.SyncClient` class gets away with this because it doesn't have any ties to a specific async event loop instance. The same cannot be said of `httpcore`, however.
What you can do is enforce your sync API to be used as a context manager which will then acquire and release a blocking portal lease from the provider.

I've considered an alternative where these event loops are global, and their shutdown is tied to the interpreter's shutdown using `atexit()`. But I'm not (yet) convinced that is a good solution either.

**Reply by agronholm** [reply]
If you don't care about asyncio compatibility, who is the target audience of your library?

**Reply by dhirschfeld** [reply]
It's for an internal company app.

If the api is being used in a sync context then the users won't know or care about the event loop it's running on in the background.

In an async context it should still work fine on `asyncio` as the users would never call the `.sync` property and so never construct a `SyncAPI` instance. So, for this use-case I think it's immaterial what backend the SyncAPI uses?

**Reply by agronholm** [reply]
If you can guarantee that it will only ever be used on Trio in an async context, then this could work, yes. I don't know exactly _why_ it works, and doing this is not something I would support, but if it works for you, then go ahead.

**Reply by agronholm** [reply]
What's the `sync` property for anyway?

**Reply by dhirschfeld** [reply]
> *What you can do is enforce your sync API to be used as a context manager*

The context manager would have to be a global one though? Users will want to use the sync api in multiple different functions throughout the app so they'll try opening the context manager multiple times and hit the same problem IIUC?

i.e. they'll try
```python
with api.sync as sync_api:
    address = sync_api.get_ipaddress()
 
...

with api.sync as sync_api:
    address = sync_api.get_ipaddress()  # Event loop is closed!
```

**Reply by agronholm** [reply]
It's hard for me to give proper advice as I don't quite understand the scope and purpose of the API you're trying to create.

**Reply by dhirschfeld** [reply]
> *What's the sync property for anyway?*

It's just a convenience method to return a `SyncAPI` instance without having to first instantiate an `AsyncAPI` instance and manually pass it in to the `SyncAPI` constructor.

In the actual app I use a `wrapt.ObjectProxy`. I kept that out of the example to avoid muddying the waters, but in practice it looks more like:

```python
from __future__ import annotations

from functools import cached_property
from ipaddress import IPv4Address

import anyio
import httpx
import trio
import wrapt
from anyio.from_thread import BlockingPortalProvider


portal_provider = BlockingPortalProvider('trio')


class AsyncAPI:
    _client: httpx.AsyncClient

    def __init__(self):
        self._client = httpx.AsyncClient()

    @cached_property
    def sync(self) -> SyncAPI:
        return SyncAPI(self)

    async def get_ipaddress(self) -> IPv4Address:
        url = "https://ifconfig.me/ip"
        response = await self._client.get(url)
        response.raise_for_status()
        return IPv4Address(response.text)


class SyncAPI(wrapt.ObjectProxy):
    def __init__(self, wrapped):
        super().__init__(wrapped)

    def get_ipaddress(self) -> IPv4Address:
        with portal_provider as portal:
            return portal.call(self.__wrapped__.get_ipaddress)
```
```python
>>> api = AsyncAPI()
>>> api.sync.get_ipaddress()
IPv4Address('<redacted>')
>>> api.sync.get_ipaddress()
IPv4Address('<redacted>')
```

**Reply by agronholm** [reply]
> > _What you can do is enforce your sync API to be used as a context manager_
> 
> The context manager would have to be a global one though? Users will want to use the sync api in multiple different functions throughout the app so they'll try opening the context manager multiple times and hit the same problem IIUC?

This will deeply involve how your application is structured. Let me just ask this: when would you ideally expect the async event loop to be terminated in your application lifecycle?

**Reply by dhirschfeld** [reply]
For the sync apps, or in an interactive notebook session I'd want the event loop to only be closed at interpreter shutdown.

I expect the api to be hit throughout the lifetime of sync apps based on different user inputs. The sync API will likely be used by less experienced programmers who would probably trip over global context managers.

It will also be used interactively in notebooks by analysts so they'll want the ability to just run it in multiple different cells.

**Reply by agronholm** [reply]
Ok, so that then warrants the use of `atexit()` to shut down the event loop. I think you need a different pattern here – one that initializes the event loop on the first use, and only shuts it down at interpreter exit.

**Comment by dhirschfeld** [comment]
Thanks for your help @agronholm. I've gone ahead and marked this as answered since I now know *why* it's happening and how it could be worked around. 

For now I'll use the trio backend as it appears to work out of the box so is the path of least resistance. I'll make sure to come back and warn against it if I live to rue that decision! 😅

**Reply by agronholm** [reply]
I must say that I have deep misgivings about the `sync` property, as it's not idempotent, and won't even use the event loop of the source async API object. I don't expect this to ever work.

**Reply by dhirschfeld** [reply]
Using a `cached_property` returns the same object each invocation so that's idempotent right?
```python
>>> id(api.sync)
123168591184256
>>> id(api.sync)
123168591184256
```

**Reply by agronholm** [reply]
That still leaves the other problem: that it's not using the same event loop as the source async API object.

**Reply by dhirschfeld** [reply]
The wrapper holds a reference to the original `AsyncAPI` object:
```python
>>> id(api)
123168618638752
>>> id(api.sync.__wrapped__)
123168618638752
>>> id(api.sync.__wrapped__)
123168618638752
```
...which is what gets passed to the `portal.call` - `portal.call(self.__wrapped__.get_ipaddress)`.


I'll fully admit that this is well above my amateur async skills/knowledge but I'm genuinely curious why it wouldn't use the same event loop if it's the same object?

**Reply by agronholm** [reply]
Because your sync API opens a blocking portal for everything. If the original async API was not created in the same event loop, then your derived sync API will try to call the async API in a different thread, so you'll essentially be running two event loop threads at the same time. This can only ever work if you ONLY use async APIs through the blocking portal.

**Reply by dhirschfeld** [reply]
Hmm, that seems a little fragile :(

I guess I need to find a way to initialize a global event loop in a sync app... at which point it seems perhaps less work to just implement the sync api :/

**Reply by agronholm** [reply]
I don't really see a problem with anything else than your `sync` property. That's what baffles me. Is it absolutely necessary to use it as a singleton?

**Reply by dhirschfeld** [reply]
No, I just don't understand how it could make a difference.

Sorry, I'm likely being dense, but are you saying he below is correct / will work?
```python
from ipaddress import IPv4Address

import httpx
from anyio.from_thread import BlockingPortalProvider


portal_provider = BlockingPortalProvider('trio')


class AsyncAPI:
    _client: httpx.AsyncClient

    def __init__(self):
        self._client = httpx.AsyncClient()

    async def get_ipaddress(self) -> IPv4Address:
        url = "https://ifconfig.me/ip"
        response = await self._client.get(url)
        response.raise_for_status()
        return IPv4Address(response.text)


class SyncAPI:
    def __init__(self, async_api: AsyncAPI):
        self._async_api = async_api

    def get_ipaddress(self) -> IPv4Address:
        with portal_provider as portal:
            return portal.call(self._async_api.get_ipaddress)



async_api = AsyncAPI()
sync_api = SyncAPI(async_api)
```

**Reply by agronholm** [reply]
Well, the most obvious problem here is what `httpx.AsyncClient()` will do when it's instantiated outside of a running event loop. Best case scenario, it defers choosing the code path until you actually try to use its API. But this snippet doesn't show how you would use both the async and sync APIs.

**Reply by dhirschfeld** [reply]
I don't think I would ever use both apis in a single app. It would either be all sync (via the blocking portal) or it would be an async app, using only the async api.

I can't imagine in an async app why you would want to use the sync api!

**Reply by agronholm** [reply, marked as answer]
If 1) `httpx.AsyncClient` defers the code path selection to the first use of the APIs, and 2) you never use both async and sync APIs from the same process, then you should be fine.

**Reply by dhirschfeld** [reply]
The only time I've mixed the apis is testing in a Jupyter notebook but I'd be happy just saying *"Don't do that"*!

Most analysts wouldn't know `autoawait` exists, so the sync api is in large part to cater for that crowd (with the minimal maintenance overhead!)

**Comment by dhirschfeld** [comment]
Thanks again for all your help understanding the nuances! It has given me a lot to think about! 😅

**Reply by agronholm** [reply]
No problem, this really helped me to understand the practical challenges of this approach, and I'll be sure to incorporate it in my talk on Wednesday.

**Comment by dhirschfeld** [comment]
***xref:*** An example using `ExitStack` - https://gist.github.com/agronholm/5a8a2773b82369195a308c910566070a

**Comment by ewjoachim** [comment]
Hello !

I seems this issue is the same as https://github.com/encode/httpcore/discussions/659 which snowballs to https://github.com/pydantic/pydantic-ai/issues/748

Just wanted GitHub to keep a link between those related issues :) Also: if the same issue keeps beeing seen downstream, does it mean comething should change in anyio ?
- Either have it work (though it makes sense if we can't take an open tcp connection and put it in another loop just like that)
- Or make a more obvious error, pointing to the error ?

I'm interested in an opinion of who did "_wrong_" (with all sorts of quotes) in the pydantic case according to anyio's knowledgeable folks: httpcore or pydantic ? (or the end user ? But I don't think so)

**Reply by agronholm** [reply]
From reading the issue it seems like they're not managing the connection pool correctly.

**Reply by ewjoachim** [reply]
Thanks a lot for weighing in :) 🙏
